serve/mlc_serve/engine/sync_engine.py

```diff
@@ -100,33 +101,6 @@ def step(self) -> InferenceStepResult:
 
         outputs = list[RequestOutput]()
 
-        # TODO: consolidate into a single function
-        for state in list(self.current_batch.values()):
-            finish_reason = None
-            if state.is_finished:
-                finish_reason = FinishReason.Stop
-            if should_stop_by_length(state, self.max_context_length):
-                finish_reason = FinishReason.Length
-
-            if finish_reason is not None:
-                outputs.append(
-                    RequestOutput(
-                        state.request_id,
-                        [
-                            SequenceOutput(
-                                i,
-                                finish_reason=finish_reason,
-                                num_generated_tokens=len(gen_seq.generated_token_ids),
-                            )
-                            for i, gen_seq in enumerate(state.generation_sequences)
-                        ],
-                        num_prompt_tokens=state.prompt_len,
-                    )
-                )
-                self.current_batch.pop(state.request_id)
-                self.cache_manager.free_request(state)
-                del self.num_sequences_per_requests[state.request_id]
```

Comment:
Here I removed empty responses the we have been sending after a request finishes. As I described in the PR description, it is incorrect to send the same `finish_reason` for all sequences. 

I believe this is the right thing to do, but it is potentially a breaking change. So I request a careful review of this change.

Reply:
Thank you for the note! Is there any need to have a different logic for sync engine here? If not, can we further consolidate it with staging engine and have a common function? Based on the quick look, seems pretty similar: https://github.com/octoml/mlc-llm/blob/63a3f19ff6d9e593d0394d5850b8a1db4b89de68/serve/mlc_serve/engine/staging_engine_worker.py#L174

Reply:
The current code is similar but buggy, so it is not relevant. 

Sync and staging engines need slightly different logic for termination check / output creation since for the latter, stopping word and length checks are done in different places (main vs worker process). So for example, if a request is terminated early due to stopping words, even after my upcoming PR the staging worker still needs to send an empty sequence with `finish_reason = Stop` due to the delay in the stopping word detection. The new code in the staging engine worker looks like this https://github.com/masahi/mlc-llm/blob/ea2a931354fe78c79dbf688f2fd5a991438d16e5/serve/mlc_serve/engine/staging_engine_worker.py#L180-L190

Reply:
Just sent https://github.com/octoml/mlc-llm/pull/124/ for the staging engine fix.